new-tutorial/L2_large_link_prediction.py

```diff
@@ -0,0 +1,361 @@
+# embeddings.
+# 
+
+
+######################################################################
+# To obtain the representations of all the nodes this tutorial uses
+# neighbor sampling as introduced in the :doc:`node classification
+# tutorial <L1_large_node_classification>`.
+# 
+# .. note::
+# 
+#    If you would like to obtain node representations without
+#    neighbor sampling during inference, please refer to this :ref:`user
+#    guide <guide-minibatch-inference>`.
+# 
+
+def inference(model, graph, node_features):
+    with torch.no_grad():
+        nodes = torch.arange(graph.number_of_nodes())
+
+        sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+        train_dataloader = dgl.dataloading.NodeDataLoader(
+            graph, torch.arange(graph.number_of_nodes()), sampler,
+            batch_size=1024,
+            shuffle=False,
+            drop_last=False,
+            num_workers=4,
+            device='cuda')
+
+        result = []
+        for input_nodes, output_nodes, bipartites in train_dataloader:
+            # feature copy from CPU to GPU takes place here
+            inputs = bipartites[0].srcdata['feat']
+            result.append(model(bipartites, inputs))
+
+        return torch.cat(result)
+
+import sklearn.metrics
+
+def evaluate(emb, label, train_nids, valid_nids, test_nids):
+    classifier = nn.Linear(emb.shape[1], label.max().item()).cuda()
+    opt = torch.optim.LBFGS(classifier.parameters())
+    
+    def compute_loss():
+        pred = classifier(emb[train_nids].cuda())
+        loss = F.cross_entropy(pred, label[train_nids].cuda())
+        return loss
+    
+    def closure():
+        loss = compute_loss()
+        opt.zero_grad()
+        loss.backward()
+        return loss
+    
+    prev_loss = float('inf')
+    for i in range(1000):
+        opt.step(closure)
+        with torch.no_grad():
+            loss = compute_loss().item()
+            if np.abs(loss - prev_loss) < 1e-4:
+                print('Converges at iteration', i)
+                break
+            else:
+                prev_loss = loss
+                
+    with torch.no_grad():
+        pred = classifier(emb.cuda()).cpu()
+        label = label
+        valid_acc = sklearn.metrics.accuracy_score(label[valid_nids].numpy(), pred[valid_nids].numpy().argmax(1))
+        test_acc = sklearn.metrics.accuracy_score(label[test_nids].numpy(), pred[test_nids].numpy().argmax(1))
+    return valid_acc, test_acc
+
+
+######################################################################
+# Defining Training Loop
+# ----------------------
+# 
+# The following initializes the model and defines the optimizer.
+# 
+
+model = Model(node_features.shape[1], 128).cuda()
+predictor = DotPredictor().cuda()
+opt = torch.optim.Adam(list(model.parameters()) + list(predictor.parameters()))
+
+
+######################################################################
+# The following is the training loop for unsupervised learning and
```

Comment:
I understand the training only use graph information but is it accurate to say it is unsupervised?

Reply:
Changed to link prediction.